### PR TITLE
fix: re-reconcile the nodeclass status

### PR DESCRIPTION
**Pull request by Vacant2333**
Fixes #N/A 

**Description**
if users omit the necessary tags and later add them, we need to reprocess the information

**How was this change tested?**

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates 
- [ ] Yes, issue opened: # 
- [x] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/controllers/nodeclass/status/ami.go [hunk 1]
```diff
@@ -42,7 +42,9 @@ func (a *AMI) Reconcile(ctx context.Context, nodeClass *v1.EC2NodeClass) (reconc
 	if len(amis) == 0 {
 		nodeClass.Status.AMIs = nil
 		nodeClass.StatusConditions().SetFalse(v1.ConditionTypeAMIsReady, "AMINotFound", "AMISelector did not match any AMIs")
-		return reconcile.Result{}, nil
+		// If users have omitted the necessary tags and later add them, we need to reprocess the information.
+		// Returning 'ok' in this case means that the nodeclass will remain in an unready state until the component is restarted.
+		return reconcile.Result{RequeueAfter: 15 * time.Second}, nil
```
Comment (jonathan-innis) [inline pkg/controllers/nodeclass/status/ami.go:47]: Yeah, this makes sense -- definitely agree that we should be re-polling. I'm wondering if we should just use the same RequeueAfter as we do in the successful case -- since there's really no reason to poll more aggressively here.

Just generally, I think we should be doing time.Minute RequeueAfter since that's our cache timeout interval so polling at that frequency can ensure that we get refreshed data more frequently. This should apply across the reconcilers -- subnet, security group, and amis

File: pkg/controllers/nodeclass/status/subnet.go [hunk 2]
```diff
@@ -40,7 +40,9 @@ func (s *Subnet) Reconcile(ctx context.Context, nodeClass *v1.EC2NodeClass) (rec
 	if len(subnets) == 0 {
 		nodeClass.Status.Subnets = nil
 		nodeClass.StatusConditions().SetFalse(v1.ConditionTypeSubnetsReady, "SubnetsNotFound", "SubnetSelector did not match any Subnets")
-		return reconcile.Result{}, nil
+		// If users have omitted the necessary tags and later add them, we need to reprocess the information.
+		// Returning 'ok' in this case means that the nodeclass will remain in an unready state until the component is restarted.
+		return reconcile.Result{RequeueAfter: 15 * time.Second}, nil
```
Comment (jonathan-innis) [inline pkg/controllers/nodeclass/status/subnet.go:45]: Should we add a test for each of these reconcilers to ensure that we don't return an empty result and instead return a RequeueAfter value?

**Review by jonathan-innis — COMMENTED** [review]
Thanks for the change! Looks good. Just a couple comments

File: pkg/controllers/nodeclass/status/readiness.go [hunk 3]
```diff
@@ -38,7 +39,9 @@ func (n Readiness) Reconcile(ctx context.Context, nodeClass *v1.EC2NodeClass) (r
 	if nodeClass.AMIFamily() == v1.AMIFamilyAL2023 {
 		if err := n.launchTemplateProvider.ResolveClusterCIDR(ctx); err != nil {
 			nodeClass.StatusConditions().SetFalse(status.ConditionReady, "NodeClassNotReady", "Failed to detect the cluster CIDR")
-			return reconcile.Result{}, fmt.Errorf("failed to detect the cluster CIDR, %w", err)
+			// If users have omitted the necessary tags and later add them, we need to reprocess the information.
+			// Returning 'ok' in this case means that the nodeclass will remain in an unready state until the component is restarted.
+			return reconcile.Result{RequeueAfter: 15 * time.Second}, fmt.Errorf("failed to detect the cluster CIDR, %w", err)
```
Comment (jmdeal) [inline pkg/controllers/nodeclass/status/readiness.go:44]: A few of things here:
- This error case isn't driven off tags, we could fail due to a networking issue or insufficient IAM permissions to discover the ClusterCIDR.
- We're already re-reconciling since we return the error, this should take care of a potential transient networking issue. I don't think requeuing faster is appropriate here (though we wouldn't anyway since the error is respected first by controller-runtime when determining when to requeue). 

I think this can be reverted to the original state.

File: pkg/controllers/nodeclass/status/ami.go [hunk 4]
```diff
@@ -42,7 +42,9 @@ func (a *AMI) Reconcile(ctx context.Context, nodeClass *v1.EC2NodeClass) (reconc
 	if len(amis) == 0 {
 		nodeClass.Status.AMIs = nil
 		nodeClass.StatusConditions().SetFalse(v1.ConditionTypeAMIsReady, "AMINotFound", "AMISelector did not match any AMIs")
-		return reconcile.Result{}, nil
+		// If users have omitted the necessary tags and later add them, we need to reprocess the information.
+		// Returning 'ok' in this case means that the nodeclass will remain in an unready state until the component is restarted.
```
Comment (jmdeal) [inline pkg/controllers/nodeclass/status/ami.go:46]: nit: let's specify that the ommited tags are on the resources, not tags ommited from the AMI selector terms. Same general suggestion applies to subnets and security groups.
```suggestion
		// If users have omitted the necessary tags from their AMIs and later add them, we need to reprocess the information.
		// Returning 'ok' in this case means that the nodeclass will remain in an unready state until the component is restarted.
```

File: pkg/controllers/nodeclass/status/readiness.go (hunk 3 above)
Comment (Vacant2333) [reply pkg/controllers/nodeclass/status/readiness.go:44]: thanks for your reminder

File: pkg/controllers/nodeclass/status/subnet.go (hunk 2 above)
Comment (Vacant2333) [reply pkg/controllers/nodeclass/status/subnet.go]: im not sure if we need the test, how do u think? if its required?

File: pkg/controllers/nodeclass/status/ami.go (hunk 1 above)
Comment (Vacant2333) [reply pkg/controllers/nodeclass/status/ami.go]: thanks for your reminder, i have check the cache timeout its 2 minues,

**Comment by Vacant2333** [comment]
@jmdeal @jonathan-innis hi, can u help me review again, i have resolve the comments, thank u!

**Comment by Vacant2333** [comment]
@jmdeal @tzneal can u help me review?

**Review by jonathan-innis — APPROVED** [review]
LGTM 🚀

**Comment by MrMarkW** [comment]
I believe a change like this should have also been applied to https://github.com/aws/karpenter-provider-aws/blob/main/pkg/controllers/nodeclass/readiness.go.
